Added option to copy blocks when editing a configuration.

RichardTysonIsis opened the pull request:
### Description of work

I added the option to copy blocks when editing a configuration. For this, I added:
- A copy button to the block tab of the config editing panel.
- A method that copies one or more blocks when they are selected in the table.
- A method to rename these blocks so that each block has a unique name.
- A keyboard listener that allows for the key combination ctrl+c to be detected and used as a shortcut to copy the selected blocks in the table.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/3293

### Acceptance criteria

- The blocks are copied correctly, are added to the configuration and are displayed in the table.
- Both the shortcut and the button work, for one or more blocks.
- The button is correctly placed.
- A unique name is given to the copied block or blocks.

### Unit tests

Unit tests added to test the view model for the Block Editor Panel.

### System tests

None required. 

### Documentation
Added a sentence explaining that blocks can be copied both with the copy button or the keyboard shortcut, so that users are aware of the keyboard shortcut, in the user manual (at https://github.com/ISISComputingGroup/ibex_user_manual/wiki/CreateandManageConfigurations).

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

RichardTysonIsis commented:
The latest commit made a few changes to the the code: 
- Added a view model for the block editor panel. This view model now contains the unique name setting logic. It also uses a method that does the same as what mine did (but better).
- Created unit tests for the view model.
- Corrected a few checkstyle errors here and there.
- Changed the logic in the some of the methods (under Dom's supervision).

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/blocks/BlocksEditorPanel.java:
```diff
@@ -142,15 +163,59 @@ public void mouseDoubleClick(MouseEvent e) {
+        
+        /*
+         * Allows for a keyboard shortcut (Ctrl+c) to copy block.
+         */
```
There is a question about consistency here. No other keyboard shortcuts exist in this view (for new block, delete block, etc). I think we should either add a complete set of keyboard shortcuts or none at all.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/editing/blocks/BlocksEditorPanel.java:
```diff
@@ -142,15 +163,59 @@ public void mouseDoubleClick(MouseEvent e) {
+                    config.addNewBlock(added);
+                } catch (DuplicateBlockNameException e1) {
+                    // TODO Auto-generated catch block
+                    e1.printStackTrace();
```
Remove `TODO` and add more appropriate error handling. At a minimum send it to the `IsisLog`, but really you probably want to bring up an error dialog so that the user notices.

RichardTysonIsis commented:
The latest commit:
- Reordered the buttons so that they are all in one row.
- Changed the naming function so that the copied blocks have valid names. I don't think it's worth using a validator as the user wouldn't be the one to choose the name so there isn't a need to display a warning.
- Added a warning dialog if the copying function tried setting a block in the config with an already used name. To test this you can just comment out the part of the copying function that sets the new name to the new block.
- Added new keyboard shortcuts to copy, add and edit blocks (there already was one to delete blocks). I've set these to `CTRL-D`, `CTRL-A` and `CTRL-E`. My only concern is that these shortcuts may not feel intuitive to the users: `CTRL-D` is for example used to delete a line in Eclipse or `CTRL-A` is usually used to "select all". I'll add a paragraph in the [documentation](https://github.com/ISISComputingGroup/ibex_user_manual/wiki/CreateandManageConfigurations) to make the shortcuts explicit but I don't know if that would help users notice them.

RichardTysonIsis commented:
The latest commit added the condition that blocks are editable for the keyboard shortcut `CTRL-E` to work as I only realised whilst editing the documentation that blocks inherited from components shouldn't be edited.
